devito/types/basic.py

```diff
@@ -317,66 +271,125 @@ def _C_typedata(self):
     def _C_ctype(self):
         return dtype_to_ctype(self.dtype)
 
+    def _subs(self, old, new, **hints):
+        """
+        This stub allows sympy.Basic.subs to operate on an expression
+        involving devito Scalars.  Ordinarily the comparisons between
+        devito subclasses of sympy types are quite strict.
+        """
+        try:
+            if old.name == self.name:
+                return new
+        except AttributeError:
+            pass
+
+        return self
+
     # Pickling support
-    _pickle_args = ['name']
-    _pickle_kwargs = ['dtype']
+    _pickle_args = []
+    _pickle_kwargs = ['name', 'dtype', 'is_const']
     __reduce_ex__ = Pickable.__reduce_ex__
 
 
-class AbstractCachedSymbol(AbstractSymbol, Cached):
+class Symbol(AbstractSymbol, Cached):
+
     """
-    Base class for dimension-free symbols, cached by both Devito and SymPy.
+    A scalar symbol, cached by both Devito and SymPy, which does not carry
+    any data.
 
-    For more information, refer to the documentation of AbstractSymbol.
+    Notes
+    -----
+    A Symbol may not be in the SymPy cache, but still be present in the
+    Devito cache. This is because SymPy caches operations, rather than
+    actual objects.
     """
 
+    @classmethod
+    def _cache_key(cls, *args, **kwargs):
```

Comment:
note for reviewers: we cache our own Symbols based on the (immutable) arguments through which they have been created. So if Devito sees the same arguments again, it will return the cached symbol instead of creating a new one. Note that w/o immutability, this would have been impossible.

Reply:
i think that answers my question in `test_caching.py`.
maybe this explanation could be added to `Notes` above